examples/blockchain/src/network.rs

```diff
@@ -233,130 +95,200 @@ impl aleph_bft::Network<NetworkData> for Network {
-                }))
-                .build()
-        };
-
-        swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
 
         let network = Network {
             msg_to_manager_tx,
             msg_from_manager_rx: msg_from_store,
         };
 
         let network_manager = NetworkManager {
-            swarm,
+            id,
+            address,
+            addresses,
+            bootnodes,
+            listener,
+            consensus_tx: msg_for_store,
             consensus_rx: msg_to_manager_rx,
+            block_tx: block_to_data_io_tx,
             block_rx: block_from_data_io_rx,
         };
 
         Ok((
-            network,
             network_manager,
+            network,
             block_from_data_io_tx,
             block_to_data_io_rx,
             msg_for_network,
             msg_from_manager,
         ))
     }
-}
 
-impl NetworkManager {
+    fn recipient_to_addresses(&self, recipient: Recipient) -> HashMap<usize, Address> {
+        let mut addr: HashMap<usize, Address> = HashMap::new();
+        match recipient {
+            Recipient::Node(n) => {
+                let n: usize = n.into();
+                assert!(n < self.addresses.len());
+                if n != self.id {
+                    if let Some(a) = &self.addresses[n] {
+                        addr.insert(n, a.clone());
+                    }
+                }
+            }
+            Recipient::Everyone => {
+                for n in 0..self.addresses.len() {
+                    if n != self.id {
+                        if let Some(a) = &self.addresses[n] {
+                            addr.insert(n, a.clone());
+                        }
+                    };
+                }
+            }
+        }
+        addr
+    }
+
+    fn reset_dns(&mut self, n: usize) {
+        if !self.bootnodes.contains_key(&(n as u32)) {
+            error!("Reseting address of node {}", n);
+            self.addresses[n] = None;
+        }
+    }
+
+    fn send(&mut self, message: Message, recipient: Recipient) {
+        let addr = self.recipient_to_addresses(recipient);
+        for (n, addr) in addr.into_iter() {
+            match self.try_send(message.clone(), &addr) {
+                Ok(_) => (),
+                Err(_) => self.reset_dns(n),
+            };
+        }
+    }
+
+    fn try_send(&self, message: Message, address: &Address) -> std::io::Result<()> {
+        debug!("Trying to send message {:?} to {:?}", message, address);
+        address.connect()?.write_all(&message.encode())
+    }
+
+    fn dns_response(&mut self, id: usize, address: Address) {
+        self.addresses[id] = Some(address.clone());
+        self.try_send(Message::DNSResponse(self.addresses.clone()), &address)
+            .unwrap_or(());
+    }
+
     pub async fn run(&mut self, mut exit: oneshot::Receiver<()>) {
+        let dns_ticker_delay = std::time::Duration::from_millis(1000);
+        let mut dns_ticker = Delay::new(dns_ticker_delay).fuse();
```

Comment:
You don't need to fuse `dns_ticker` and `dns_hello_ticker`. In the loop you mutate them anyway, to a new future. Also maybe `timeout` would be better than ticker?.

Reply:
Also you could use `tokio::time::interval` here if you would like to use a fancy way for intervals

Reply:
fixed